File: 304-range-sum-query-2d-immutable/304-range-sum-query-2d-immutable.cpp
class NumMatrix {
public:
    vector<vector<int>>dp;
    NumMatrix(vector<vector<int>>& matrix) {
        int m =  matrix.size();
        int n = matrix[0].size();
        vector<int>v(n+1,0);
        dp.push_back(v);
        for(int i=1;i<=m;i++){
            for(int j=1;j<=n;j++){
                v[j] = matrix[i-1][j-1]+v[j-1]+dp[i-1][j]-dp[i-1][j-1];
            }
            dp.push_back(v);
        } 
    }
    
    int sumRegion(int r1, int c1, int r2, int c2) {
        r2++; c2++;
        return dp[r2][c2]-dp[r1][c2]-dp[r2][c1]+dp[r1][c1];
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */

File: 304-range-sum-query-2d-immutable/304-range-sum-query-2d-immutable.py
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        m = len(matrix)
        n = len(matrix[0])
        self.sum=[[0] * (n+1) for _ in range(m+1)]
        for i in range(1,m+1):
            for j in range(1,n+1):
                self.sum[i][j] = matrix[i-1][j-1]+self.sum[i-1][j]+self.sum[i][j-1]-self.sum[i-1][j-1]

    def sumRegion(self, r1: int, c1: int, r2: int, c2: int) -> int:
        r2+=1
        c2+=1
        return self.sum[r2][c2]+self.sum[r1][c1]-self.sum[r2][c1]-self.sum[r1][c2]


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)